### PR TITLE
Redact Ingest Processor

### PR DESCRIPTION
**Pull request by davidkyle**
The redact processor uses the Grok rules engine to redact matching Grok patterns. Existing patterns from the [Grok pattern bank](https://github.com/elastic/elasticsearch/tree/main/libs/grok/src/main/resources/patterns/ecs-v1) can be referenced directly and new patterns added inline in the `pattern_definitions` option. 

One application of the redact processor is to obscure Personal Identifying Information by configuring the processor to detect known patterns such as email or IP addresses. 

In an ingest pipeline the redact processor could be augmented by a Named Entity Recognition model to detect and remove names, places. 

```
PUT _ingest/pipeline/redact
{
  "processors": [
    {
      "redact": {
        "field": "to_redact",
        "patterns": ["%{EMAILADDRESS:EMAIL}", "%{IP:IP_ADDRESS}", "%{CREDIT_CARD:CREDIT_CARD}"],
        "pattern_definitions": {
          "CREDIT_CARD": "\\d{4}[ -]\\d{4}[ -]\\d{4}[ -]\\d{4}"
        }
      }
    }
  ]
}
```

Given an input document with the field `to_redact`
```
{
  "to_redact": test@elastic.co sent an email from the IP 192.168.0.1."
}
```

The redact processor, as configured above, will emit
```
{
  "to_redact": <EMAIL> sent an email from the IP <IP_ADDRESS>."
}
```

The matched text is replaced by the Grok pattern name. The `<` and `>` tokens surrounding the replaced text are configurable via the `prefix` and `suffix` options.

### CODE REVIEW

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/IngestCommonPlugin.java [hunk 1]
```diff
@@ -86,7 +86,8 @@ public Map<String, Processor.Factory> getProcessors(Processor.Parameters paramet
             entry(NetworkDirectionProcessor.TYPE, new NetworkDirectionProcessor.Factory(parameters.scriptService)),
             entry(CommunityIdProcessor.TYPE, new CommunityIdProcessor.Factory()),
             entry(FingerprintProcessor.TYPE, new FingerprintProcessor.Factory()),
-            entry(RegisteredDomainProcessor.TYPE, new RegisteredDomainProcessor.Factory())
+            entry(RegisteredDomainProcessor.TYPE, new RegisteredDomainProcessor.Factory()),
+            entry(RedactProcessor.TYPE, new RedactProcessor.Factory(createGrokThreadWatchdog(parameters)))
```
Comment (joegallo) [inline modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/IngestCommonPlugin.java:90]: I'm not sure whether we should be invoking `createGrokThreadWatchdog` twice, or merely creating a single value once and passing it to both the `grok` and `redact` processor factories. I'll figure out the answer and let you know.

Comment (davidkyle) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/IngestCommonPlugin.java]: Good catch! The `MatchWatchdog` is used by all the Grok processors on this node and all the `Grok`s that those processors create, it makes sense to use the same watchdog for the `Grok`s created by the redact processor. 

I pushed  https://github.com/elastic/elasticsearch/pull/92951/commits/940c3c1cdcd94473cb09f3a2e4e8efcd77229528

**Comment by joegallo** [comment]
Generally I think this looks great, please take my comments as relatively minor notes along the way towards me getting to a ✅.

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RedactProcessor.java [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+     * @param redactedEndToken Matched and redacted regions are ended with this token
+     * @return The redacted text
+     */
+    static String matchRedact(String fieldValue, List<Grok> groks, String redactedStartToken, String redactedEndToken) {
+        byte[] utf8Bytes = fieldValue.getBytes(StandardCharsets.UTF_8);
+
+        RegionTrackingMatchExtractor extractor = new RegionTrackingMatchExtractor();
+        for (var grok : groks) {
+            String className = grok.captureConfig().get(0).name();
+            extractor.setCurrentClass(className);
```
[the remark this answers is not in the document]
Comment (davidkyle) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/RedactProcessor.java]: `patternName` is better, I've used that consistently throughout now

File: docs/reference/ingest/processors/redact.asciidoc [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+}
+--------------------------------------------------
+// TESTRESPONSE[s/2023-02-01T16:53:14.560005377Z/$body.docs.0.doc._ingest.timestamp/]
+
+[[redact-custom-patterns]]
+==== Custom Patterns
+
+If one of the existing Grok https://github.com/elastic/elasticsearch/blob/{branch}/libs/grok/src/main/resources/patterns/ecs-v1[patterns]
+does not fit your requirements extend the patterns with the `pattern_definitions` option.
+New patterns can be defined with a regular expression or combine
```
[the remark this answers is not in the document]
Comment (davidkyle) [reply docs/reference/ingest/processors/redact.asciidoc]: Updated with an example of defining a custom pattern

File: docs/reference/ingest/processors/redact.asciidoc [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+  ]
+}
+--------------------------------------------------
+// TESTRESPONSE[s/2023-02-01T16:53:14.560005377Z/$body.docs.0.doc._ingest.timestamp/]
+
+[[grok-watchdog-redact]]
+==== Grok watchdog
+
+The watchdog interrupts expressions that take too long to execute.
+When interrupted the Redact processor fails with an error.
```
Comment (abdonpijpelink) [inline docs/reference/ingest/processors/redact.asciidoc:228]:
```suggestion
When interrupted, the Redact processor fails with an error.
```

**Review by abdonpijpelink — APPROVED** [review]
LGTM from a docs perspective. I've left some minor suggestions.

**Review by joegallo — APPROVED** [review]
Great feature! Thanks for writing it.

**Pull request merged**